### PR TITLE
[Tune] Enable experiment restore from moved cloud uri

### PR DESCRIPTION
**Pull request by justinvyu**
## Why are these changes needed?

This PR fixes experiment restoration from a different cloud URI to save future results and checkpoints to new URI instead of continuing to write to the old location. The workflow of starting a local experiment, uploading the experiment dir to cloud, then restoring from the URI from a different cluster is also possible now.

Example:
```python
# initial experiment experiment directory at "s3://original/exp_name"
# experiment directory moved to "s3://new/new_exp_name"
Tuner.restore("s3://new/new_exp_name")
# new results/checkpoints should get synced to the new URI "s3://new/new_exp_name"
```

This is a follow-up to https://github.com/ray-project/ray/pull/29920, which enabled experiment restoration from a moved local experiment directory.

I also took the chance to simplify some of the trial loading/serialization logic.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tune/execution/trial_runner.py [hunk 1]
```diff
@@ -845,19 +791,34 @@ def resume(
...
+            # Ex: moved local/cloud experiment directory
+            trial.local_dir = self._local_checkpoint_dir
+            trial.sync_config = self._sync_config
+            trial.experiment_dir_name = self._experiment_dir_name
+
+            # Avoid creating logdir in client mode for returned trial results,
+            # since the dir might not be creatable locally.
+            # TODO(ekl) this is kind of a hack.
+            if not ray.util.client.ray.is_connected():
+                trial.init_logdir()  # Create logdir if it does not exist
```
Comment (justinvyu) [inline python/ray/tune/execution/trial_runner.py:816]: I moved this out of `Trial.__setstate__`. Are trials serialized/deserialized in any situation that would require this to stay within `__setstate__`? My understanding is that trial objects don't get shipped around and stay on the Tune driver. Only serialized/deserialized on experiment checkpoint and restore, which is handled here.

File: python/ray/tune/execution/trial_runner.py [hunk 2]
```diff
@@ -585,6 +526,12 @@ def search_alg(self):
     def scheduler_alg(self):
         return self._scheduler_alg
 
+    @property
+    def _remote_checkpoint_dir(self):
+        if self._sync_config.upload_dir and self._experiment_dir_name:
+            return os.path.join(self._sync_config.upload_dir, self._experiment_dir_name)
```
Comment (Yard1) [inline python/ray/tune/execution/trial_runner.py:532]: Does this work for paths starting with eg. "s3://"?

File: python/ray/tune/execution/trial_runner.py (hunk 1 above)
Comment (Yard1) [reply python/ray/tune/execution/trial_runner.py:817]: What's the reason for moving this?

File: python/ray/tune/analysis/experiment_analysis.py [hunk 3]
```diff
@@ -143,13 +140,10 @@ def _load_checkpoints_from_latest(self, latest_checkpoint: List[str]) -> None:
 
             if "checkpoints" not in experiment_state:
                 raise TuneError("Experiment state invalid; no checkpoints found.")
+
             self._checkpoints_and_paths += [
-                (_decode_checkpoint_from_experiment_state(cp), Path(path).parent)
```
Comment (krfricke) [inline python/ray/tune/analysis/experiment_analysis.py:147]: Should we remove the function? I think it's not used anywhere else

File: python/ray/tune/execution/trial_runner.py (hunk 1 above)
Comment (krfricke) [reply python/ray/tune/execution/trial_runner.py:817]: You're correct @justinvyu and I think your solution is much cleaner. For setstate/getstate, it's ok to just restore exactly the state that was saved. If properties are overwritten, that should happen in the function that issues the restore. No need for magic here.

File: python/ray/tune/tests/test_tuner_restore.py [hunk 4]
```diff
@@ -746,6 +757,66 @@ def test_tuner_restore_from_moved_experiment_path(
     assert not old_local_dir.exists()
 
 
+def test_tuner_restore_from_moved_cloud_uri(
+    ray_start_2_cpus, tmp_path, clear_memory_filesys
+):
+    """Test that that restoring an experiment that was moved to a new remote URI
```
Comment (krfricke) [inline python/ray/tune/tests/test_tuner_restore.py:763]: nit
```suggestion
    """Test that restoring an experiment that was moved to a new remote URI
```

File: python/ray/tune/tests/test_tuner_restore.py [hunk 5]
```diff
@@ -746,6 +757,66 @@ def test_tuner_restore_from_moved_experiment_path(
...
+        "memory:///original/exp_dir", str(tmp_path / "moved" / "new_exp_dir")
+    )
+    delete_at_uri("memory:///original")
+    upload_to_uri(str(tmp_path / "moved"), "memory:///moved")
+
+    tuner = Tuner.restore("memory:///moved/new_exp_dir", resume_errored=True)
+    # This is needed because the mock memory:/// filesystem doesn't
+    # sync checkpoints properly, so this just copies the original checkpoint
+    # to the new local directory.
+    # NOTE: A new local directory is used since the experiment name got modified.
```
Comment (krfricke) [inline python/ray/tune/tests/test_tuner_restore.py:793]: Can you elaborate what it doesn't do?
Is this just specific to the memory fs or does this come up with other fs as well?

**Review by krfricke — APPROVED** [review]
This is great, thanks! Only a couple of nits

File: python/ray/tune/execution/trial_runner.py (hunk 1 above)
Comment (Yard1) [reply python/ray/tune/execution/trial_runner.py:817]: @krfricke doesn't this make trials effectively unserializable in the general case unless the trainable is registered? I don't think that's an issue, but perhaps something to consider

File: python/ray/tune/tests/test_tuner_restore.py (hunk 5 above)
Comment (justinvyu) [reply python/ray/tune/tests/test_tuner_restore.py:793]: Just memory fs. I think the problem is that the trainable actor uploads its checkpoints to a different `memory://` filesystem that's not accessible by the main process. Then, when I download from the URI, only files uploaded by the driver are there.

**Pull request merged**